src/Compilers/CSharp/Portable/Binder/Lifetime.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// A representation of the program region in which a 'ref' can be used,
+    /// which is limited to what is expressible in C#.
+    /// </summary>
+    /// <remarks>
+    /// For example, in this design, all lifetimes have a known relationship to all other lifetimes.
+    /// </remarks>
+    internal readonly struct Lifetime
+    {
+        private const uint CallingMethodRaw = 0;
+        private const uint ReturnOnlyRaw = 1;
+        private const uint CurrentMethodRaw = 2;
+
+        /// <summary>
+        /// For the purpose of escape verification we operate with the depth of local scopes.
+        /// The depth is a uint, with smaller number representing shallower/wider scopes.
+        /// Since sibling scopes do not intersect and a value cannot escape from one to another without
+        /// escaping to a wider scope, we can use simple depth numbering without ambiguity.
+        /// </summary>
+        private readonly uint _value;
+        private Lifetime(uint value) => _value = value;
+
+        /// <summary>
+        /// The "calling method" scope that is outside of the containing method/lambda.
+        /// If something can escape to this scope, it can escape to any scope in a given method through a ref parameter or return.
+        /// </summary>
+        public static Lifetime CallingMethod => new Lifetime(CallingMethodRaw);
+
+        /// <summary>
+        /// The "return-only" scope that is outside of the containing method/lambda.
+        /// If something can escape to this scope, it can escape to any scope in a given method or can be returned, but it can't escape through a ref parameter.
+        /// </summary>
+        public static Lifetime ReturnOnly => new Lifetime(ReturnOnlyRaw);
+
+        /// <summary>
+        /// The "current method" scope that is just inside the containing method/lambda.
+        /// If something can escape to this scope, it can escape to any scope in a given method, but cannot be returned.
+        /// </summary>
+        public static Lifetime CurrentMethod => new Lifetime(CurrentMethodRaw);
+
+        /// <summary>
+        /// Gets a lifetime which is "empty". i.e. which refers to a variable whose storage is never allocated.
+        /// </summary>
+        public static Lifetime Empty => new Lifetime(uint.MaxValue);
+
+        /// <summary>
+        /// Gets a lifetime which is narrower than the given lifetime.
+        /// Used to "enter" a nested local scope.
+        /// </summary>
+        public Lifetime Narrower()
+        {
+            var result = new Lifetime(this._value + 1);
+            // Narrower() operator should always result in a local lifetime
```

Comment:
What if we apply `Narrower` to `CallingMethod`? Then we'd get `ReturnOnly` which is returnable

Reply:
Have adjusted the asserts to be in terms of the raw value.

I made the asserts intolerant of calling Narrower() on `Lifetime.CallingMethod`, for example, because we don't have a reason to do that.